packages/headless-react/src/search-engine.tsx

```diff
@@ -0,0 +1,108 @@
+import {Controller, CoreEngine, SearchEngine} from '@coveo/headless';
+import {
+  ControllerDefinitionsMap,
+  InferControllerFromDefinition,
+  SearchEngineDefinitionOptions,
+  defineSearchEngine as defineBaseSearchEngine,
+  mapObject,
+} from '@coveo/headless/ssr';
+import {
+  Context,
+  createContext,
+  useContext,
+  useCallback,
+  useMemo,
+  useSyncExternalStore,
+} from 'react';
+import {
+  ContextCSRState,
+  ContextState,
+  ControllerHook,
+  InferControllerHooksMapFromDefinition,
+  ReactSearchEngineDefinition,
+} from './types';
+
+export function capitalize(s: string) {
+  return `${s.slice(0, 1).toUpperCase()}${s.slice(1)}`;
+}
+
+function isCSRContext<
+  TEngine extends CoreEngine,
+  TControllers extends ControllerDefinitionsMap<TEngine, Controller>,
+>(
+  ctx: Exclude<ContextState<TEngine, TControllers>, null>
```

Comment:
Can you please explain the reasoning behind?

Reply:
My reasoning is very subjective of course. I believe that given your experience, your judgement would be better than mine on this.

Here's my reasoning though:
# Reason 1: it's more intuitive
Instinctively, when I see a type like this, I assume it's most likely an interface, and `null` isn't assignable to an interface.

# Reason 2: less presumptuous
I usually think of `null` or `undefined` as meaning "not specified" or "optional". Usually, whether something is optional depends on where the type is needed in the first place.

For example, if we had a type like `Person` and a method called `setParent(child, newParent)`, it contextually makes sense that `child` isn't nullable/optional but `newParent` is nullable/optional.

# Reason 3: it's easier to extend
To make something non-optional into something optional, we can either use unions or the `someParam?: SomeType` syntax.

To make something optional into something non-optional, we can either use `Exclude<SomeType, null | undefined>` or `Required<ParentOfSomeType>`.

Making something non-optional into something optional seems easier/prettier to me than the other way around.

# Counter point
Arguably that `ContextState` is very strictly tied to the React context. Although we assert that it's not null, in reality, it can be null until that assertion.

Additionally, we could argue that `ContextState` should technically be something like `type ContextState = SSRState | CSRState | null`.

Reply:
Nm, this is obsolete now with the refactor to use singleton context